Begin work on reducing Red's potential bad API calls

mikeshardmind opened the pull request:
### Type

- [ ] Bugfix
- [X] Enhancement
- [ ] New feature

### Description of the changes

Adds a new exception type for our use. 
Changes the potential exceptions raised by `context.send` (Breaking)
Logs unavoidable 403s raised by `context.send`, `context.tick`, `context.react_quietly`

see <https://discordapp.com/channels/81384788765712384/381871767846780928/610143340238733322> (discord API server announcement) for details on why now on this.

Even without this though, we should be reducing the bad calls we might make.

Commit:
Begin work on reducing Red's potential bad API calls

Commit:
breaking notice addition

Commit:
add a small bit of handling to core

mikeshardmind commented:
This should be good for review. I added a little bit to help cog devs track it down if running in debug. Beyond that, if this exception goes unhandled, it really should do nothing since we've already determined we'd be trying to be doing something we don't have perms for.

Rapptz commented:
This seems like a waste. Jake and I have discussed this heavily internally and you're more likely to get banned by 429s than 403s when sending messages which makes the permission check rather strange and inconsistent with other send(...) methods for no good reason.  Not to mention the library handles the 429 case already so you're not going to get banned.

Everyone who got banned from 403 spams got it from spamming audit logs. If you want to investigate 403s then do so there IMO.

mikeshardmind commented:
> Everyone who got banned from 403 spams got it from spamming audit logs. If you want to investigate 403s then do so there IMO.

This is an initial part of making the entire bot look before leap with regard to discord's API. Anything we can do to reduce the API impact of Red is a good thing. The only reason I started with context is because it also helps out with 3rd party code which may not be as well written as the core bot.

Just because the example case which caused this announcement to happen was a bot spamming audit logs without perms doesn't mean other pathological cases shouldn't be prevented.

> This seems like a waste. Jake and I have discussed this heavily internally and you're more likely to get banned by 429s than 403s when sending messages which makes the permission check rather strange and inconsistent with other send(...) methods for no good reason. 

The official announcement is pretty clear. There's no stated difference between the bad calls. you get only so many in an hour. If for some reason the official announcement is incorrect and you have knowledge of that, I'd suggest you talk with Jake again and have him clarify it.

DevilXD commented:
Could this exception inherit from `discord.Forbidden`? Since it's infeasible to track 429 or 401 here, this exception will only be used to check for correct permissions before trying to do the action, and both - trying to do the action and failing because of missing permissions - and - trying to do the action and failing on the permission check - should cause a `discord.Forbidden` (or some subclass of it ofc) to be raised. This would also make it a non-breaking change.

mikeshardmind commented:
I don't really like the idea of inheriting from discord.Forbidden. It has attributes which we should not be populating, and removing those attributes but faking the subclass is a liskov violation. Even if we were to populate them, we would need to populate them with a compatible type for it to be compatible. An example of some of these attributes is actually visible in the PR, currently being used to log unavoidable 403s separately so that bot owners have a means to handle someone causing those frequently (may be enhanced with some automatic things later as well)

Commit:
found a potential auditlog issue too

Commit:
Possible solution to the exception type

I'm not super happy about this manner of handling it, but if we want to
inherit from `discord.Forbidden`, this is how I'd go about it to avoid
the Lisvok violation.

mikeshardmind commented:
Following up on Danny's comment about the audit logs, I actually found a place where it's possible we hit the audit logs with 403s, though the conditions are a little contrived. It's been addressed in a commit here.

Following up with the exception type, I've gone ahead and put (in it's own commit for easy removal) what I think that should look like if we absolutely insist on inheriting from discord.Forbidden here, but I'm not entirely convinced doing so is a good idea. This includes manual creation of a ClientResponse object

Method for manually creating the ClientResponse object safely is a slight variation on what aiohttp itself uses for it's tests.

Commit:
Revert exception type changes

This commit can be removed prior to merge if for some reason we want
that back

This reverts commit d6edaeb12e76fe86cbbf8016aa11a7fb20f63528.

mikeshardmind commented:
Exception type has been reverted.

Per discussion in discord, we're gonna hope we can get some more info prior to 3.2

There has been no official assurance made that 403s which occur as a result of malicious actors (eg. user blocking bot, using command which reacts) wont cause issues.

Without such assurances, it would not be difficult to have 17 bad requests per second caused by a handful of userbots if someone wanted to take down a public instance (as of the enhanced limits)

However, with **official** assurances that this is not how discord will handle it, we can simplify a lot of this. 

This is on hold for now, but should be included in 3.2 is we don't have those assurances by then.

mikeshardmind commented:
There's a comment from Discord Jake in the reddit thread which was linked in that announcement and assurances that they will handle the abuse case should it arise as an issue. 

With that in mind, I'm closing this. Other changes may still be prudent, but large changes here shouldn't be needed.

#2944 seems like a less drastic middle ground on addressing the issue, though it too might be unneeded.
